### DISCUSSION TITLE
Appendix Heading: Numbering

### DISCUSSION THREAD

**Discussion by ansgarlichter** [typst/typst#3630] [Q&A] [answered]
I am currently designing my appendix for my thesis. I want to have the first-level-heading in the appendix to have "Appendix A:" as numbering, such in the following example:

![image](https://github.com/typst/typst/assets/43957906/a60f6066-f769-4b79-87d0-23b3dbaebf36)

Unfortunately the references do not work as expected. I want the reference to not include the ":" part. The default numbering does this for example with the "." if I set the numbering to "1.1.". How can I achieve this?
My current approach has 2 problems:
- [ ] The ":" is included in the reference. It is "Appendix A:" but it should be "Appendix A"
- [ ] The reference starts with a space. It is " Appendix A" but it should be "Appendix A". I do not have an idea where this space originates from.

Do I have to write a custom ref function or is there some other way? As I think the default numbering function does not include the "." if the numbering is "1.1." I think my numbering function may contain a mistake.

This is my current Typst code:
```typst
#counter(heading).update(0)
#set heading(
  numbering: (..nums) => {
    let vals = nums.pos()
    let currentPosition = vals.at(0)
    let numberingTopHeading = "ABCDEFGHIJ".at(vals.at(0) - 1)
    if vals.len() == 1 {
      return "Appendix " + numberingTopHeading + ":"
    }
    else {
      return numberingTopHeading + "." + nums.pos().slice(1).map(str).join(".")
    }
  },
  supplement: (heading) => {
    if heading.depth == 1 {
      return ""  
    } else {
      return "Appendix"
    }
  }
)

#show ref: it => {
  let eq = math.equation
  let el = it.element
  if el != none and el.func() == eq {
    // Override equation references.
    numbering(el.numbering, ..counter(eq).at(el.location()))
  } else {
    // Other references as usual.
    it
  }
}

= Literature Research <app:ch:literatureResearch>
#lorem(10)

= Expert Interviews <app:ch:expertInterviews>
#lorem(10)

== Guideline <app:sec:expertInterviews:guideline>
#lorem(10) @app:sec:expertInterviews:guideline shows #lorem(10) and @app:ch:literatureResearch shows blub
```

The final document will have also many headings with the default numbering "1.1.".

I have tried to solve this with this custom ref function:
```typst
#show ref: it => {
  let eq = heading
  let el = it.element
  if el != none and el.func() == eq and type(el.numbering) == function and el.supplement != [Appendix] {
    let str = numbering(
      el.numbering,
      ..counter(eq).at(el.location())
    )

    str.slice(0, str.len() - 1)  
  } else {
    // Other references as usual.
    it
  }
}
```

Is this really the best and cleanest way to achieve this? Do you know a better alternative?

*Reproduction URL*
https://typst.app/project/rM_0Zvb50cJ7_YGFDdtYSf

**Comment by laurmaedje** [comment]
I would suggest leaving the colon out of the numbering and instead adding it in a heading show rule. That will sidestep the issue and also seems semantically more correct to me.

**Reply by ansgarlichter** [reply]
Thank you for pointing me into the right direction. I implemented it as following:

```typst
#counter(heading).update(0)
#set heading(
  numbering: (..nums) => {
    let vals = nums.pos()
    let currentPosition = vals.at(0)
    let numberingTopHeading = "ABCDEFGHIJ".at(vals.at(0) - 1)
    if vals.len() == 1 {
      return "Appendix " + numberingTopHeading
    }
    else {
      return numberingTopHeading + "." + nums.pos().slice(1).map(str).join(".")
    }
  },
  supplement: (heading) => {
    if heading.level == 1 {
      return ""  
    } else {
      return "Appendix"
    }
  }
)
#show heading: it => {
  if it.numbering != none and it.level == 1 {
    counter(heading).display() + ":"
    h(0.3em)
  } else if it.numbering != none {
    counter(heading).display() + "."
    h(0.3em)
  }

  it.body
  parbreak()
}
```

This looks more clean and works as expected.
Nevertheless, the problem with the space in front of the headings in the appendix which belong to the first level still persits. Do you have any idea where this originates and how I can fix this?

The following code leads to a double space between `and` and `Appendix ...`:
```typst
and @app:ch:literatureResearch
```
![image](https://github.com/typst/typst/assets/43957906/50d5ccf6-e8ea-4710-af91-4450847cb83a)

If I change this to the following it works, but is hard to ensure through the whole thesis:
```typst
and@app:ch:literatureResearch
```
![image](https://github.com/typst/typst/assets/43957906/270a7b28-ef3c-4238-9801-9280635af3f2)

See also: [https://typst.app/project/roh4szf36uz8idLKORiP0W](https://typst.app/project/roh4szf36uz8idLKORiP0W)

**Reply by laurmaedje** [reply, marked as answer]
The `ref` show rule will automatically add a space between the supplement and the numbering, unless it detects that the supplement is empty. In this case though it is an empty string, which it detects as a text element and thus not empty. You can change `return ""` to just `return` or `return []` to fix it.

Beyond that, you rarely need to use return in Typst at all because all expressions in a block are joined up into a final output value. Making use of this, a simplified but equivalent version of your heading set rules could be this:
```typ
#set heading(
  numbering: (..nums) => {
    let vals = nums.pos()
    if vals.len() == 1 {
      [Appendix ]
    }
    numbering("A.1", ..vals)
  },
  supplement: (heading) => if heading.level != 1 [Appendix],
)
```
However, looking at this, I only just really realized what you want to do. 

I think the "Appendix" word shouldn't be part of the numbering, but also of the show rule. Would something like this would work?
```typ
#set heading(numbering: "A.1", supplement: [Appendix])
#show heading: it => {
  if it.level == 1 and it.numbering != none {
    [#it.supplement #counter(heading).display():]
  } else if it.numbering != none {
    [#counter(heading).display().]
  }

  h(0.3em)
  it.body
  parbreak()
}
```

**Reply by ansgarlichter** [reply]
Thanks for your explanation. I learned very much about Typst and how great show rules are :) 

I tried your approach. This works as expected for the heading. Nevertheless, the Table of Contents doesn't include the "Appendix" part. Therefore I included `Appendix` in the numbering to influence both parts.  Is there a another show rule with which I can influence the TOC?

I also could just do a separate TOC for the Appendix to solve to problem if there is no such show rule.

**Reply by laurmaedje** [reply]
You could adjust the TOC via a show rule on `outline.entry`. By default it doesn't show the supplement.

**Reply by ansgarlichter** [reply]
I will give this a try. I marked your answer as I am confident that this will work. Thank you very much!
Just as a feedback: I think more more emphasis on the show rules in the documentation may help. Don't know if others hat similar problems.
